File: day_06/main.py
import copy
import sys

def traverse(grid):
    is_cycle = False
    def check_cycle(x, y, direction):
        match(direction):
            case '>':
                # check if going right results in cycle
                if y < len(grid) -1 and grid[x][y+1] == "R":
                    return True
            case '<':
                # check if going left results in cycle
                if y > 0 and grid[x][y-1] == "L":
                    return True
            case '^':
                # check if going up results in cycle
                if x > 0 and grid[x-1][y] == "U":
                    return True
            case 'v':
                # check if going down results in cycle
                if x < len(grid[x]) -1 and grid[x+1][y] == "D":
                    return True


    def move_guard(x, y, direction):
        is_cycle = False
        direction_map = {'^': 'U', 'v': 'D', '>': 'R', '<': 'L'}
        grid[x][y] = direction_map.get(direction, '')
        if direction == '>':  # Move right
            if y < len(grid[x]) - 1:
                if grid[x][y + 1] == "#" and x < len(grid) - 1:
                    grid[x + 1][y] = 'v'  # Go down
                    is_cycle = check_cycle(x+1, y, 'v')
                else:
                    grid[x][y + 1] = '>'
                    is_cycle = check_cycle(x, y+1, '>')
                return (True, is_cycle)
        elif direction == '<':  # Move left
            if y > 0:
                if grid[x][y - 1] == "#" and x > 0:
                    grid[x - 1][y] = '^'  # Go up
                    is_cycle = check_cycle(x-1, y, '^')
                else:
                    grid[x][y - 1] = '<'
                    is_cycle = check_cycle(x, y-1, '<')
                return (True, is_cycle)
        elif direction == '^':  # Move up
            if x > 0:
                if grid[x - 1][y] == "#" and y < len(grid[x]) - 1:
                    grid[x][y + 1] = '>'  # Go right
                    is_cycle = check_cycle(x, y+1, '>')
                else:
                    grid[x - 1][y] = '^'
                    is_cycle = check_cycle(x-1, y, '^')
                return (True, is_cycle)
        elif direction == 'v':  # Move down
            if x < len(grid) - 1:
                if grid[x + 1][y] == "#" and y > 0:
                    grid[x][y - 1] = '<'  # Go left
                    is_cycle = check_cycle(x, y-1, '<')
                else:
                    grid[x + 1][y] = 'v'
                    is_cycle = check_cycle(x+1, y, 'v')
                return (True, is_cycle)
        return (False, is_cycle)

    while True:
        still_here = False
        is_cycle = False
        for x, row in enumerate(grid):
            for y, cell in enumerate(row):
                if cell in ('>', '<', '^', 'v'):
                    still_here, is_cycle = move_guard(x, y, cell)
                    break
        if not still_here:
            # print("finished")
            # positions_visited = sum(cell in ('U', 'D', 'R', 'L') for row in grid for cell in row)
            # print(positions_visited)
            # print(grid)
            return False
        if is_cycle:
            return True

def main(input):
    file_name = input + ".txt"

    if input == "input":
        width = 130
        height = 130
    elif input == "example":
        width = 10
        height = 10

    x = 0
    y = 0
    grid = [[0 for w in range(width)] for h in range(height)] 

    with open(file_name, "r") as file:
        for line in file:
            for char in line:
                if (char == '\n' or y >= width or x >= height):
                    continue
                grid[x][y] = char
                y += 1
            y = 0
            x += 1
    count = 0

    for x, row in enumerate(grid):
        for y, cell in enumerate(row):
            print(x, y)
            if cell not in ('#', '<', '>', 'v', '^'):
                temp = grid[x][y]
                grid[x][y] = '#'
                is_cycle = traverse(copy.deepcopy(grid))
                if is_cycle:
                    count += 1
                grid[x][y] = temp
    print(count)
if __name__ == "__main__":
    input = sys.argv[1]
    main(input)

File: day_06/main.c
#include <stdio.h>
#include <stdlib.h>
#include<stdbool.h>
#include<string.h>

#define MAX_HEIGHT 130
#define MAX_WIDTH 130

typedef struct {
    int x;
    int y;
} Point;

typedef struct {
    int modX;
    int modY;
} Modifiers;

char grid[MAX_HEIGHT][MAX_WIDTH];

int visited[MAX_HEIGHT][MAX_WIDTH]; // 0: unvisited, 1-4: direction
int visitedTemp[MAX_HEIGHT][MAX_WIDTH];

Point guardPos = {-1, -1};
char direction;
bool still_here = true;
int cycles = 0;

bool isGuard(char c) {
    return (c == '<' || c == 'v' || c == '^' || c == '>');
}

int charToDir(char c) {
    switch (c) {
        case '<': return 1; // Left
        case 'v': return 2; // Down
        case '>': return 3; // Right
        case '^': return 4; // Up
        default: return 0; // Invalid
    }
}

bool withinBounds(Modifiers modifiers) {
    return (0 <= (guardPos.x + modifiers.modX) && (guardPos.x + modifiers.modX) < MAX_WIDTH) &&
        (0 <= (guardPos.y + modifiers.modY) && (guardPos.y + modifiers.modY) < MAX_HEIGHT);
}

bool isObstruction(Modifiers modifiers) {
    return grid[guardPos.x + modifiers.modX][guardPos.y + modifiers.modY] == '#';
}

bool isCycle() {
    return charToDir(direction) == visitedTemp[guardPos.x][guardPos.y];
}

char turningRight(char direction) {
    switch (direction) {
        case '<':
            return '^';
        case '>':
            return 'v';
        case '^':
            return '>';
        case 'v':
            return '<';
        default:
            break;
    }
    return 0;
}

Modifiers getModifiers() {
    Modifiers modifiers = {0, 0};
    switch(direction) {
        case '<':
            modifiers.modX = 0;
            modifiers.modY = -1;
            break;
        case '>':
            modifiers.modX = 0;
            modifiers.modY = +1;
            break;
        case '^':
            modifiers.modX = -1;
            modifiers.modY = 0;
            break;
        case 'v':
            modifiers.modX = +1;
            modifiers.modY = 0;
            break;
        default:
            break;
    }
    return modifiers;
}

int move(Modifiers modifiers) {
    if (withinBounds(modifiers)) {
        if (!isObstruction(modifiers)) {
            visitedTemp[guardPos.x][guardPos.y] = charToDir(direction);
            guardPos.x += modifiers.modX;
            guardPos.y += modifiers.modY;
        } else {
            direction = turningRight(direction);
        }
    } else {
        still_here = false;
        visitedTemp[guardPos.x][guardPos.y] = charToDir(direction);
        guardPos.x = -1;
        guardPos.y = -1;
    }
    return 0;
}

int traverse() {
    // initial direction
    direction = grid[guardPos.x][guardPos.y];
    while (still_here) {
        if (isGuard(direction)) {
            if (!isCycle()) {
                move(getModifiers());
            } else {
                cycles++;
                return 0;
            }
        } else {
            still_here = false;
        }
    }

    return 0;
}

int main() {
    FILE *file = fopen("input.txt", "r");
    if (file == NULL) {
        printf("Error: Could not open file \n");
        exit(1);
    }

    int x = 0;
    int y = 0;

    char c;
    while ((c = fgetc(file)) != EOF) {
        if (y < MAX_WIDTH) {
            if (isGuard(c)) {
                grid[x][y] = c;
                guardPos.x = x;
                guardPos.y = y;
            } else if (c == '#') {
                grid[x][y] = c;
            }
            y++;
        } else {
            y = 0;
            x++;
        }   
    }

    Point initialPos = guardPos;

    if (guardPos.x != -1 && guardPos.y != -1) {
        traverse();
    }

    int count = 0;
    for (int i = 0; i < MAX_HEIGHT; i++) {
        for (int j = 0; j < MAX_WIDTH; j++) {
            visited[i][j] = visitedTemp[i][j];
            if (visited[i][j] > 0) {
                count++;
            }
        }
    }

    for (int i = 0; i < MAX_HEIGHT; i++) {
        for (int j = 0; j < MAX_WIDTH; j++) {
            if (visited[i][j] > 0) {
                char temp = grid[i][j];
                grid[i][j] = '#';
                guardPos = initialPos;
                still_here = true;
                memset(visitedTemp, 0, sizeof(visitedTemp));
                traverse();
                grid[i][j] = temp;
            }
        }
    }

    printf("count: %d\n", count);
    printf("cycles: %d\n", cycles);

    fclose(file);
    return 0;
}
